web/html/src/manager/maintenance/maintenance-windows.js

```diff
@@ -0,0 +1,249 @@
+const SpaRenderer  = require("core/spa/spa-renderer").default;
+
+const messagesCounterLimit = 1;
+const hashUrlRegex = /^#\/([^\/]*)(?:\/(.+))?$/;
+
+
+function getHashId() {
+    const match = window.location.hash.match(hashUrlRegex);
+    return match ? match[2] : undefined;
+}
+
+function getHashAction() {
+    const match = window.location.hash.match(hashUrlRegex);
+    return match ? match[1] : undefined;
+}
+
+class MaintenanceWindows extends React.Component {
+
+    constructor(props) {
+        super(props);
+
+        ["delete", "handleForwardAction", "handleDetailsAction", "handleEditAction", "handleResponseError",
+            "update", "refreshCalendar"]
+            .forEach(method => this[method] = this[method].bind(this));
+        this.state = {
+            type: type,
+            messages: [],
+            schedules: [],
+        };
+    }
+
+    componentDidMount() {
+        this.updateView(getHashAction(), getHashId());
+        this.state.type === "schedule" && this.getCalendarNames();
+        window.addEventListener("popstate", () => {
+            this.updateView(getHashAction(), getHashId());
+        });
+    }
+
+    updateView(action, id) {
+        if (action === "details" && id) {
+            this.getDetails(id, "details");
+        } else if (id || !action) {
+            this.handleForwardAction();
+        } else {
+            this.setState({action: action});
+        }
+        this.clearMessages();
+    }
+
+    getMaintenanceSchedules = () => {
+        const endpoint = "/rhn/manager/api/maintenance/" + this.state.type + "/list";
+        return Network.get(endpoint, "application/json").promise
+            .then(schedules => {
+                this.setState({
+                    action: undefined,
+                    selected: undefined,
+                    schedules: schedules
+                });
+            }).catch(this.handleResponseError);
+    };
+
+    getCalendarNames = () => {
+        const endpoint = "/rhn/manager/api/maintenance/calendar";
+        return Network.get(endpoint, "application/json").promise
+            .then(calendarNames => {
+                this.setState({
+                    calendarNames: Array.from(Array(calendarNames.length).keys()).map(id => ({id: Number(id), text: calendarNames[id]}))
+                });
+            }).catch(this.handleResponseError);
+    };
+
+    getDetails(row, action) {
+        const endpoint = "/rhn/manager/api/maintenance/" + this.state.type + "/" + row + "/details";
+        return Network.get(endpoint, "application/json").promise
+            .then(item => {
+                this.setState({
+                    selected: item,
+                    action: action
+                });
+                history.pushState(null, null, "#/" + action + "/" +
+                    (this.state.type === "schedule" ? item.scheduleId : item.calendarId));
+            }).catch(this.handleResponseError);
+    }
```

Comment:
What about to extract the fetching data from the rest of the page? Maybe you can create a `maintenance-windows-api.js` where to collect all the api enpoint calls to fetch data. In case we need to fetch the same data from anywhere else, this could be reused easily.